### ISSUE TITLE
Packet Terraform provider does not register / save server state for next-available reserved instances

### ISSUE THREAD

**Issue by taylor** [terraform-providers/terraform-provider-packet#133]
When provisioning Packet servers with Terraform using next-available option for reserved instances the servers are provisioned but no state information is saved. Running terraform again will create a new set of servers. Terraform destroy will say there are zero servers to destroy.

### Terraform information
Terraform version: 0.11.10
Packet provider plugin: 1.2.5 (~> 1.2)
Affected Resource: packet_device

**Terraform Configuration Files**
_packet.tf_
```
resource "packet_device" "cnfs" {
  count            = "${ var.packet_node_count }"
  hostname         = "${ var.name }${ count.index + 1 }"
  facility         = "${ var.packet_facility }"
  project_id       = "${ var.packet_project_id }"
  plan             = "${ var.packet_master_device_plan }" # Packet.net machine type
  billing_cycle    = "${ var.packet_billing_cycle }" # hourly or monthly
  operating_system = "${ var.packet_operating_system }"
 
  connection {
    type     = "ssh"
    user     = "root"
    private_key = "${ file("/root/.ssh/id_rsa")}"
}
  
  provisioner "ansible" {
    plays {
      playbook ={
        file_path = "${ var.playbook }"
      }
    }
  }
}
```

_variables.tf_
```
provider "packet" {
  auth_token = "${ var.packet_api_key }"
}
variable "name" { default = "pktgen" }
variable "packet_node_count" { default = "1" }
variable "packet_project_id" {}
variable "packet_api_key" {}
variable "packet_facility" { default = "sjc1" }
variable "packet_billing_cycle" { default = "hourly" }

# VM Image and size
variable "packet_master_device_plan" { default = "m2.xlarge.x86" }
variable "packet_operating_system" { default = "ubuntu_16_04" }

#Ansible
variable "playbook" {}
```

_override_
```
# Provision L2 Networking 

resource "packet_device" "cnfs" {
  hardware_reservation_id = "next-available"
}
```

### Debug Output
n/a

### Panic Output
n/a

### Expected Behavior
What should have happened?
- Terraform state for newly created resources should be saved
- Terraform destroy should delete resources that were created
- Terraform apply should create new resources when they already exist

### Actual Behavior
What actually happened?
- Terraform state not saved
- Terraform destroy, destroys 0 resources
- Terraform apply creates new resources when existing ones are still running

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Update config with Packet account/project info
1. `terraform apply`

### References
- https://github.com/cncf/cnf-testbed/issues/215

/cc @denverwilliams

**Comment by t0mk** [comment]
hi @taylor thanks for reporting. I will take a look on this tomorrow.

**Comment by taylor** [comment]
Sounds good @t0mk.  FYI, we think it could be related to a previously fixed bug: https://github.com/packethost/packngo/issues/117

**Comment by t0mk** [comment]
hi @taylor, is this the issue you mentioned in the Community channel?

```
* packet_device.cnfs[0]: 1 error(s) occurred:

* packet_device.cnfs.0: json: cannot unmarshal string into Go struct field PortData.bonded of type bool
* packet_device.cnfs[1]: 1 error(s) occurred:

* packet_device.cnfs.1: json: cannot unmarshal string into Go struct field PortData.bonded of type bool
```
I am not sure if the deserialization of PortData.bonded is related to hardware reservations. Do you mean that when the `apply` run blows on the unmarshal error, the tf state is not stored, but the devices on the hardware reservations are created? That's pretty bad I guess. The content of the state files is not produced by the provider, but by the TF logic. There is a way to store the state partially from the provider though, which might alleviate this issue.

I've just tried to provision a (only one) server with your setup on reserved hardware (different plan though) and it worked OK for me.

You did well to fill the issue template, but I'd like to ask for an additional info - If you can, please upload terraform log from the apply run (when it errs). You can get it as

```bash
TF_LOG=DEBUG terraform apply &> log
```

**Comment by taylor** [comment]
@t0mk we do get this error for our k8s deployments.  We also saw the same problem w/o this error.  This only happens with the reserved instances in our account.

**Comment by denverwilliams** [comment]
@t0mk here is a copy of the terraform log for the m2.xlarge reserved servers.
https://gist.githubusercontent.com/denverwilliams/c15c1102a8a10ad387024f23712566ce/raw/28eefc4934ea284c3b3bd03043b051ed16aaa7cd/terrform.debug

**Comment by t0mk** [comment]
@denverwilliams @taylor Thanks for the log. I see the problem, it's an API type mismatch. I reported it to Packet folks. 

I checked if this could be alleviated with the Partial state, but it seems not, SetPartial is only good for Update functions..

**Comment by t0mk** [comment]
@denverwilliams @taylor I heard this was fixed in the Packet API. Could you please test again, and close this issue if the `terraform apply` goes through, and your devices are created properly?

**Comment by rata** [comment]
I've used the "next-available" feature with terraform provider 1.7.2 and it works fine now. As far as I can see, there is no change to be made on packet terraform provider but it was fixed in the Packet API.

To confirm, when using "next-available" I can see that the tfstate file is updated with the reservation id (something like "111-sadasd-2sd") for the instance and no changes are required by terraform on later plans/applies. Furthermore, I couldn't reproduce none of the issues mentioned here, IMHO it really seems they are fixed in Packet API now
